### PR TITLE
fish-like (Fuzzy) completion

### PR DESCRIPTION
**Pull request by GuillaumeLeclerc**
For the moment I applied the fuzzy search to the files only.

Applying it to python completions was leading to too many matches.

But maybe we could apply it too git completions or commands.

Tell me what you think.

_Complexity is `O(n)` for each possible completion_

### CODE REVIEW

File: xonsh/completer.py [hunk 1]
```diff
@@ -84,13 +84,23 @@ def _path_from_partial_string(inp, pos=None):
 
 WS = set(' \t\r\n')
 
+def fuzzyMatch(ref, typed):
```
Comment (scopatz) [inline xonsh/completer.py]: PEP8 function names

File: xonsh/completer.py [hunk 2]
```diff
@@ -391,14 +402,19 @@ def path_complete(self, prefix, start, end, cdpath=False):
         """Completes based on a path name."""
         tilde = '~'
         paths = set()
+        sep = get_sep()
+        prefix_parts = prefix.split(sep)
+        prefix_parts[-1] = ""
+        search_pattern = sep.join(prefix_parts)
```
Comment (scopatz) [inline xonsh/completer.py]: I think this could be better achieved with an rpartition than a split.

**Comment by scopatz** [comment]
Thanks a million for putting this in, @GuillaumeLecler! Could you please put in an entry to the CHANGELOG and some tests. Also making this PEP8 compliant is important too.  Thanks!

**Comment by GuillaumeLeclerc** [comment]
I know PEP8 is important but I come (and do on a daily basis) languages that have very different coding conventions. It's hard for me to do the switch multiple times a day. I will change them and try to be more careful.

File: xonsh/completer.py (hunk 2 above)
Comment (GuillaumeLeclerc) [inline xonsh/completer.py]: I don't know about this. I will investigate and use it.

Comment (scopatz) [inline xonsh/completer.py]: Strings have an rpartition method, see https://docs.python.org/3/library/stdtypes.html?highlight=rpartition#str.rpartition

**Comment by funkyfuture** [comment]
it's very handy to review and fix code styling with the atom editor and [a linter](https://atom.io/packages/linter-flake8) and its enabled error panel.

**Comment by GuillaumeLeclerc** [comment]
I don't want to work outside of vim.

But it might be cool to have a linter integrated in the build process. I usually do it in my other projects.

I even make it part of the CI tests to be sure the code base keep the same code style event if contributors use a different editor.

**Comment by santagada** [comment]
I think this pull request might have become stale (because @adqm already implemented it and merged) but on the good news is that we have a linter in the build process now :).

I think this can be closed.

**Comment by scopatz** [comment]
@GuillaumeLeclerc please let us know your thoughts on this.  Otherwise we'll close it in a few days. Thanks!

**Comment by GuillaumeLeclerc** [comment]
Hello,

I was very busy but I'm back. I will install the latest version of xonsh and tell you if it adds more features than @adqm PR.

**Comment by GuillaumeLeclerc** [comment]
The implementation on master branch is correct (and works better than mine on dotfiles). It's safe to close.

**Pull request closed without merging**